app/services/invoice_renderer.rb

```diff
@@ -0,0 +1,29 @@
+class InvoiceRenderer
+  def render_to_string(order)
+    renderer.render_to_string(args(order))
+  end
+
+  def args(order)
```

Comment:
this method is not used outside the class, right? in that case, it's better if we move it to private to keep the public API to the minimum.

Reply:
The OrdersController calls this:
```
  def print
    render InvoiceRenderer.new.args(@order)
  end
```

I tried to move the `render` call into the InvoiceRenderer but that failed badly.

```
RuntimeError (ActionController::RackDelegation#content_type= delegated to @_response.content_type=, but @_response is nil: #<ApplicationController:0x005630be1fca30 @_routes=nil, @_action_has_layout=true, @_headers={"Content-Type"=>"text/html"}, @_status=200, @_request=nil, @_response=nil, @_config={}, @_prefixes=["application"], @_lookup_context=#<ActionView::LookupContext:0x005630be144340 @details_key=#<ActionView::LookupContext::DetailsKey:0x005630c488b0f8 @hash=-3205369017263900504>, @details=...
```

I think `render` tries to store variables like the content type in the request object of the controller. But in the service we just have a new ApplicationController without any request object stored in there. `render` is only to be called on a the active controller that actually received a request. This is another example why internal state is bad.

Reply:
I know there is a way to move render outside the controller although I never used it but I don't think we need it here. I was just talking about the `#args` method itself but I see what you did now :+1: